Make authentication type optional when api id is used

anastyn opened the pull request:
Made authenticationType attribute optional if user provided the apiId attribute, since it is not used anyway.
It will allow users to create cleaner configs if all they want is to update resources in an existing AppSync API.

Also added 3 more ignored attributes to the README.
And added a log highlighting the same information if user provided an API ID.

Moved the check for the appId attribute to addResources method to make it easier to see from the code where it is used and which attributes are ignored. The ignored attributes are all attributes that modify AppSync API settings.

Updated the unit tests to make sure that the ignored resources are actually ignored.
Tested the creation of a new API and the updates to an existing API.

The attributes are ignored (instead of throwing an exception) due to the backwards compatibility with the previous version, where user could provide the apiId and call the delete operation without removing existing attributes. Otherwise I would add an exception if any of them was provided together with apiId.

anastyn commented:
Any comments?

@bboure maybe some suggestions?

The general idea here is that when the appId is used Serverless maintains only data sources, schema, and resolvers. AppSync configuration is maintained separately in the existing API endpoint settings.

bboure commented:
LGTM @anastyn Thank you again.

Can we address the missing schema issue also mentioned in #395? If the schema is not present, we just don't deploy the `AWS::AppSync::GraphQLSchema` resource. 

The problem is that this is a very uncommon use case. I'd rather keep an error if no schema is present for the sake of saving most people time and headaches if they just forget to set it.

Can we implement the same logic as for authentication?
- if `apiId` is present, ignore the missing schema
- if `apiId` is **not** present. Throw an error and stop the deployment process.

In fact, this last one is missing I think. I don't see any validation about that. I think it just fails because it tries to read a file that doesn't exist. Can we add a proper validation explaining the issue?

Thank you!

anastyn commented:
I think we could do that.
But it looks like it's more like a feature than the fix implemented in this pull request.
Can we merge this one, and discuss and review the changes to the schema in another pull request?

bboure commented on src/index.js:
```diff
@@ -514,7 +532,7 @@ class ServerlessAppsyncPlugin {
           Properties: {
             ApiCachingBehavior: config.caching.behavior,
-            ApiId: config.apiId || { 'Fn::GetAtt': [logicalIdGraphQLApi, 'ApiId'] },
+            ApiId: { 'Fn::GetAtt': [logicalIdGraphQLApi, 'ApiId'] },
```
@anastyn Why did you remove support for Api Keys and Caching?
As far as I can tell, someone might want to re-use an existing API but still deploy new API keys or configure caching.
This is definitely something possible and that was possible before this change.
I think it is useful to keep support and also not making any breaking change.

anastyn commented:
@bboure I've just ignored all AppSync settings for simplicity. I think it makes sense to still manage the caching via Serverless, I've added it back. Please check.

For ApiKeys it's a bit more complicated. The authentication type is not modified and not needed when the ApiId is provided, and the existing API may have any authentication type, for example Cognito or OpenID, something different than ApiKey authentication type. 

Even if you add authenticationType attribute to Serverless config it will not update It in AppSync when apiId is provided. So I've removed the keys for simplicity.
The existing AppSync API could be created from any CF template or manually so we can't get its authenticationType attribute from the template.

anastyn commented:
What we can do is to add a condition and allow modifying apiKeys ignoring the provided authentication type since it is not used anyway, as it was before. Just maybe put some info about it in the doc and in the log.

bboure commented:
@anastyn Thank you. 
Let's keep it simple for now.
On a side note, I recently came across [this article](https://aws.amazon.com/blogs/aws/new-import-existing-resources-into-a-cloudformation-stack/).
It looks like it can solve most of the issues we're trying to solve with the `apiId` attribute.
I tried it and it works pretty well, although I did not try it with an AppSync stack. With all the resources involved (Api, Schema, resolvers, data-sources), it might become more tedious and complex to do but I think it's doable. 
If you have som time, maybe have a try with it and see if it works in your particular use-case.

bboure commented:
I resolved the conflit and I will merge now.

Thank you!

anastyn commented:
@bboure Interesting idea, I'll take a look. But it seems that AppSync resources are not supported at this time.
Here is the full list of the supported resources:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import-supported-resources.html

bboure commented:
Yep, I figured that out later 😞 
Let's keep an eye on it. For now, your PR will do.
Thanks again
